calculator_test.go

```diff
@@ -24,31 +24,84 @@ func TestSubtract(t *testing.T) {
 }
 
 type TestCase struct {
-	A    float64
-	B    float64
+	name string
+	a, b float64
+	f    func(float64, float64) float64
 	want float64
 }
 
-func TestMultiply(t *testing.T) {
+func TestCalculator(t *testing.T) {
 	t.Parallel()
 	testCases := []TestCase{
 		{
-			A:    2,
-			B:    2,
+			name: "Add two positive integers",
+			a:    2, b: 2,
+			f:    calculator.Add,
 			want: 4,
 		},
 		{
-			A:    -3,
-			B:    3,
-			want: -9,
+			name: "Add two negative integers",
+			a:    -2, b: -2,
+			f:    calculator.Add,
+			want: -4,
+		},
+		{
+			name: "Add a positive integer and a negative fraction ",
+			a:    5, b: -2.48,
+			f:    calculator.Add,
+			want: 2.52,
+		},
+		{
+			name: "Subtract a larger integer from a smaller integer",
+			a:    3, b: 5,
+			f:    calculator.Subtract,
+			want: -2,
+		},
+		{
+			name: "Subtract a negative integer from a negative integer",
+			a:    -9, b: -5,
+			f:    calculator.Subtract,
+			want: -4,
+		},
+		{
+			name: "Multiply two positive integers",
+			a:    4, b: 20,
+			f:    calculator.Multiply,
+			want: 80,
+		},
+		{
+			name: "Multiply a positive fraction by a negative fraction",
+			a:    4, b: -9,
+			f:    calculator.Multiply,
+			want: -36,
 		},
 	}
 
 	for _, tc := range testCases {
-		got := calculator.Multiply(tc.A, tc.B)
-		if tc.want != got {
-			t.Errorf("Multiply(%f, %f): want %f, got %f", 2.0, 2.0, tc.want, got)
-		}
+		t.Run(tc.name, func(t *testing.T) {
+			got := tc.f(tc.a, tc.b)
+			if tc.want != got {
+				t.Errorf("want %f, got %f", tc.want, got)
+			}
+		})
+	}
+}
+
+type TestCaseDivide struct {
+	name        string
+	a, b        float64
+	errExpected bool
+}
+
+func TestDivide(t *testing.T) {
+	tc := TestCaseDivide{
+		a: 5, b: 0,
+		errExpected: true,
 	}
 
+	_, err := calculator.Divide(tc.a, tc.b)
+
+	if err == nil {
```

Comment:
Let's think about how we can build on this to incorporate some _valid_ (that is, non-error) test cases. We obviously can't fail the test if `err == nil` for every case; for valid cases, it's _supposed_ to be nil.

Suppose we first of all construct a boolean value that tells us whether we _received_ an error. Then we can compare that with the boolean `tc.errExpected`, which tells us whether we _wanted_ an error. They should be the same, shouldn't they? Either both false, or both true. If there's a mismatch, that should fail the test.

Only if the error status is correct should we go on to compare the data value with expectation, in just the same way that you have with the other tests.